### DISCUSSION TITLE
Wand/Delphi: Transparent outside clipping path

### DISCUSSION THREAD

**Discussion by HeinziTuberkel** [ImageMagick/ImageMagick#3385] [Help]
Hi,
I'm developing with Delphi, using IM for several progams.
Until now I've been using IM in the command line version and called it with "ShellExec". The problem is, that from ShellExec it's difficult to evaluate Results from -identify etc.
I found [https://github.com/viniciusfbb/image-magick](https://github.com/viniciusfbb/image-magick) (THANKS!) and am beginning to understand, how to use it.

I want to remove the background (make it transparent) for pictures with a clipping path and save them as .png file.
The input files can be .psd or .jpg with a clipping path from photoshop.
On the command line I used to call 
`convert <infile.psd>[0] -alpha transparent -clip -alpha opaque -channel Alpha -negate <outfile.png>`
I don't understand, how I can issue those commands/options with the Magick Wand. 
I found the funcions `MagickClipImage` and `MagickNegateImage`, but I don't know:
> 1. how to set the image alpha mode.
> 2. how to detect, if the image hase more than one frame - I suppose there must be some section in `identify`, but I can't spot it.
> 3. how to work with only one frame of the image (usually [0] - the first frame).

**Comment by snibgo** [comment]
I suggest you familiarise yourself with the wand interface, especially magick-image.c.

> 1. how to set the image alpha mode.

Try MagickSetImageAlphaChannel().

> 2. how to detect, if the image hase more than one frame - I suppose there must be some section in identify, but I can't spot it.

ImageMagick has no concept of "frames".

MagickGetNumberImages() tells you how many images are in a wand.

> 3. how to work with only one frame of the image (usually [0] - the first frame).

You can read just one image from a file, by suffixing the filename with "[0]" or whatever. Or you can read the entire file, with all its images, into one wand, then loop through the images in the wand.

**Reply by HeinziTuberkel** [reply]
> MagickSetImageAlphaChannel:

Looks like the Delphi implementaiton I'm using is not complete i.e. does not contain all functions of the wand. I have now started to extend it.
I looked into the declaration of `MagickSetImageAlphaChannel`. It does not allow the same parameters as the command line magick.
On the command line `-alpha` allows a parameter `transparent` that is not in the declaration of `MagickSetImageAlphaChannel`.

> ImageMagick has no concept of "frames".

Well, then I guess I used the wrong word for it. Is it **layers** then? or **parts**? Some Images, when I convert them to png (on the command line) are output to several .png images. To make sure, only the first contained _instance_ in the original image is converted, I always add `[0]` to the input file name. 

Anyway, thanks for your hint for just suffixing the file name with [0] when loading from file into the wand.

**Comment by fmw42** [comment]
For novices, see

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/command-line-options.php
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdown
https://imagemagick.org/script/porting.php#cli

**Comment by snibgo** [comment]
The function is:
```
MagickBooleanType MagickSetImageAlphaChannel(MagickWand *wand,
  const AlphaChannelOption alpha_type)
```
The comments say alpha_type is ActivateAlphaChannel, DeactivateAlphaChannel, OpaqueAlphaChannel, or SetAlphaChannel.

However, the type AlphaChannelOption is defined in magickcore\image.h, and has many other possible values, such as TransparentAlphaChannel.

IM deal with files and images. PNG is a file format. PNG files contain only one image. A PSD file can contain many images.
